File: 백준/Gold/1753. 최단경로/최단경로.py
import sys
import heapq
input = sys.stdin.readline
INF = int(1e9)

V, E = map(int, input().split()) # 정점의 개수, 간선의 개수
K = int(input()) # 시작 정점

graph = [[] for _ in range(V+1)]
dist = [INF] * (V+1) # 최단 경로 저장 (초기값은 무한대)

for _ in range(E):
    u, v, w = map(int, input().split())
    graph[u].append((v, w)) # 정점 u에서 v로 가는 가중치 w (여러 개 간선 존재 O)
    
pq = []
heapq.heappush(pq, (0, K)) # (거리, 정점) 우선순위큐에 삽입 (순서 주의, 거리 기준으로 자동 정렬)
dist[K] = 0 # 시작점 자신 0 출력

while pq:
    cur_dist, cur = heapq.heappop(pq) # 현재까지 가장 최단 거리 정점 꺼내기

    if dist[cur] < cur_dist: # 현재 거리보다 더 최단 거리가 있으면 컷
        continue

    for next, weight in graph[cur]: # 인접 노드의 정점, 가중치 확인
        if cur_dist + weight < dist[next]: # 인접 노드에 저장된 최단 거리보다 짧으면
            dist[next] = cur_dist + weight # 최단 거리 갱신
            heapq.heappush(pq, (cur_dist + weight, next))

for i in range(1, V+1): # 출력
    print(dist[i] if dist[i] != INF else "INF")

File: 백준/Gold/1753. 최단경로/최단경로.java
import java.util.*;
import java.io.*;

public class Main {
    static final int INF = Integer.MAX_VALUE;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int V = Integer.parseInt(st.nextToken());
        int E = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(br.readLine());

        List<int[]>[] graph = new ArrayList[V + 1];
        for (int i = 1; i <= V; i++) graph[i] = new ArrayList<>();

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());
            graph[u].add(new int[]{v, w});
        }

        int[] result = solution(V, E, K, graph);

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= V; i++) {
            if (result[i] == INF) sb.append("INF\n");
            else sb.append(result[i]).append("\n");
        }
        System.out.print(sb);
    }

    public static int[] solution(int V, int E, int K, List<int[]>[] graph) {
        int[] dist = new int[V + 1];
        Arrays.fill(dist, INF);

        // (현재 노드, 거리) -> 거리순으로 정렬
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> Integer.compare(a[1], b[1]));
        pq.offer(new int[]{K, 0});
        dist[K] = 0;

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int cur_node = cur[0];
            int cur_dist = cur[1];

            if (cur_dist > dist[cur_node]) continue;

            for (int[] next: graph[cur_node]) {
                int next_node = next[0];
                int next_dist = next[1];

                int cost = dist[cur_node] + next_dist;

                if (cost < dist[next_node]) {
                    dist[next_node] = cost;
                    pq.offer(new int[]{next_node, dist[next_node]});
                }
            }
        }
        
        
        return dist;
    }
}